irohad/ametsuchi/impl/wsv_restorer_impl.cpp

```diff
@@ -17,30 +16,44 @@ namespace iroha {
         boost::optional<std::unique_ptr<iroha::LedgerState>>,
         std::string>
     WsvRestorerImpl::restoreWsv(Storage &storage) {
-      auto block_query = storage.getBlockQuery();
-      if (not block_query) {
-        return expected::makeError("Cannot create BlockQuery");
-      }
-
-      // get all blocks starting from the genesis
-      std::vector<std::shared_ptr<shared_model::interface::Block>> blocks;
-      auto top_height = block_query->getTopBlockHeight();
-      for (decltype(top_height) i = 1; i <= top_height; ++i) {
-        auto block_result = block_query->getBlock(i);
-
-        if (auto e = boost::get<expected::Error<std::string>>(&block_result)) {
-          return expected::makeError(std::move(e)->error);
+      auto mutable_storage_result =
+          storage.createMutableStorageWithoutBlockStorage();
+      return mutable_storage_result | [&storage](auto &&mutable_storage)
+                 -> iroha::expected::Result<
+                     boost::optional<std::unique_ptr<iroha::LedgerState>>,
+                     std::string> {
+        auto block_query = storage.getBlockQuery();
+        if (not block_query) {
+          return expected::makeError("Cannot create BlockQuery");
         }
 
-        auto &block = boost::get<expected::Value<
-            std::unique_ptr<shared_model::interface::Block>>>(block_result)
-                          .value;
-        blocks.push_back(std::move(block));
-      }
-
-      storage.reset();
-
-      return storage.insertBlocks(blocks);
+        return storage.resetWsv() |
+                   [&storage, &mutable_storage, &block_query]() mutable
+               -> iroha::expected::Result<
+                   boost::optional<std::unique_ptr<iroha::LedgerState>>,
+                   std::string> {
+          // apply all blocks starting from the genesis
+          auto top_height = block_query->getTopBlockHeight();
+          for (decltype(top_height) i = 1; i <= top_height; ++i) {
+            auto block_result = block_query->getBlock(i);
```

Comment:
It seems the process will be more straight forward if we create a lambda for fetching and the insertion.